codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java

```diff
@@ -600,4 +578,107 @@ private void verifyGoAwayNotReceived() throws Http2Exception {
             }
         }
     }
+
+    private final class PrefaceFrameListener implements Http2FrameListener {
+        /**
+         * Verifies that the HTTP/2 connection preface has been received from the remote endpoint.
+         * It is possible that the current call to
+         * {@link Http2FrameReader#readFrame(ChannelHandlerContext, ByteBuf, Http2FrameListener)} will have multiple
+         * frames to dispatch. So it may be OK for this class to get legitimate frames for the first readFrame.
+         */
+        private void verifyPrefaceReceived() throws Http2Exception {
```

Comment:
I think you can get rid of the `if` check altogether and always throw.  Maybe rename this to `throwPrefaceException`?